### PR TITLE
[cli]: default web.bundler to metro for new projects

### PR DESCRIPTION
**Pull request by marklawlor**
# Why

Schema PR: https://github.com/expo/universe/pull/14308

Fulfils part of the requirement of ENG-11084. SDK 50 projects with no `web.bundler` set, and which do not have `@expo/webpack-config` installed will default to `metro`.

This changes the error message to

```
CommandError: It looks like you're trying to use web support but don't have the required dependencies installed.

Please install react-native-web@~0.19.6, react-dom@18.2.0 by running:

npx expo install react-native-web@~0.19.6 react-dom@18.2.0

If you're not using web, please ensure you remove the "web" string from the platforms array in the project Expo
config.
```

I"ll update this message in a new PR. Ideally it should just prompt to user to run `npx expo install`, but that command does not check web platform dependencies.

# Test Plan

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/@expo/cli/src/start/server/platformBundlers.ts [hunk 1]
```diff
@@ -1,15 +1,28 @@
...
+export function getPlatformBundlers(
+  projectRoot: string,
+  exp: Partial<ExpoConfig>
+): PlatformBundlers {
+  /**
+   * SDK 50+: The web bundler is dynamic based upon the presence of the `@expo/webpack-config` package.
+   */
+  let web = exp.web?.bundler;
+  if (!web) {
+    const resolved = resolveFrom.silent(projectRoot, '@expo/webpack-config/package.json');
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/platformBundlers.ts:17]: interesting. be sure to update the schema description in universe to indicate how the new default value works.

File: packages/@expo/cli/src/start/server/metro/__tests__/MetroBundlerDevServer-test.ts [hunk 2]
```diff
@@ -52,7 +52,7 @@ beforeEach(() => {
 async function getStartedDevServer(options: Partial<BundlerStartOptions> = {}) {
   const devServer = new MetroBundlerDevServer(
     '/',
-    getPlatformBundlers({ web: { bundler: 'metro' } })
+    getPlatformBundlers(__dirname, { web: { bundler: 'metro' } })
```
Comment (marklawlor) [inline packages/@expo/cli/src/start/server/metro/__tests__/MetroBundlerDevServer-test.ts:55]: This is incorrect, but the `projectRoot` value is not used as the `web.bundler` is specified.

File: packages/@expo/cli/src/export/__tests__/resolveOptions-test.ts [hunk 3]
```diff
@@ -20,12 +20,6 @@ describe(resolveOptionsAsync, () => {
     );
   });
 
-  it(`asserts not-configured platform`, async () => {
-    await expect(resolveOptionsAsync('/', { '--platform': ['web'] })).rejects.toThrow(
-      /^Platform "web" is not configured to use the Metro bundler in the project Expo config\./
-    );
-  });
-
```
Comment (marklawlor) [inline packages/@expo/cli/src/export/__tests__/resolveOptions-test.ts:28]: I removed this test as I'm unsure of its functionality. `web` is not configured to use Metro, so this is no longer expected to fail?

File: packages/@expo/cli/src/export/__tests__/resolveOptions-test.ts [hunk 4]
```diff
@@ -46,7 +40,7 @@ describe(resolveOptionsAsync, () => {
     await expect(
       resolveOptionsAsync('/', { '--platform': ['android', 'all'] })
     ).resolves.toMatchObject({
-      platforms: ['android', 'ios'],
+      platforms: ['android', 'ios', 'web'],
```
Comment (marklawlor) [inline packages/@expo/cli/src/export/__tests__/resolveOptions-test.ts:43]: As `web` uses Metro, do we expect `all` to expand to `web`?

Comment (EvanBacon) [reply packages/@expo/cli/src/export/__tests__/resolveOptions-test.ts:43]: no, because `web` is an out of tree platform and requires extra installation.

Comment (EvanBacon) [reply packages/@expo/cli/src/export/__tests__/resolveOptions-test.ts:43]: oh I see, this is for export––yes it should include all metro platforms.

File: packages/@expo/cli/src/start/server/metro/__tests__/MetroBundlerDevServer-test.ts (hunk 2 above)
Comment (EvanBacon) [reply packages/@expo/cli/src/start/server/metro/__tests__/MetroBundlerDevServer-test.ts:55]: This just needs to align with what's sent to the getConfig method, which is mocked out. A "/" would probably be less confusing.

File: packages/@expo/cli/src/start/server/platformBundlers.ts (hunk 1 above)
Comment (marklawlor) [reply packages/@expo/cli/src/start/server/platformBundlers.ts:17]: https://github.com/expo/universe/pull/14308

File: packages/@expo/cli/src/start/server/metro/withMetroMultiPlatform.ts [hunk 5]
```diff
@@ -481,7 +484,9 @@ export async function withMetroMultiPlatformAsync(
   await setupNodeExternals(projectRoot);
 
   let expoConfigPlatforms = Object.entries(platformBundlers)
-    .filter(([, bundler]) => bundler === 'metro')
+    .filter(
+      ([platform, bundler]) => bundler === 'metro' && exp.platforms?.includes(platform as Platform)
```
Comment (marklawlor) [inline packages/@expo/cli/src/start/server/metro/withMetroMultiPlatform.ts:488]: This new guard is required as `expoConfigPlatforms` included `web` for non-web projects.

File: packages/@expo/cli/src/start/server/middleware/ManifestMiddleware.ts [hunk 6]
```diff
@@ -360,8 +360,8 @@ export abstract class ManifestMiddleware<
   /** Exposed for testing. */
   async checkBrowserRequestAsync(req: ServerRequest, res: ServerResponse, next: ServerNext) {
     // Read the config
-    const bundlers = getPlatformBundlers(this.initialProjectConfig.exp);
-    if (bundlers.web === 'metro') {
+    const bundlers = getPlatformBundlers(this.projectRoot, this.initialProjectConfig.exp);
+    if (bundlers.web === 'metro' && this.initialProjectConfig.exp.platforms?.includes('web')) {
```
Comment (marklawlor) [inline packages/@expo/cli/src/start/server/middleware/ManifestMiddleware.ts:364]: This new guard is required as non-web projects still have the `bundler.web === 'metro'`

**Review by EvanBacon — APPROVED** [review]

**Pull request merged**